packages/node/src/policy-util.js

```diff
@@ -11,46 +13,120 @@ import nodeFs from 'node:fs'
 import nodePath from 'node:path'
 import * as constants from './constants.js'
 import { readJsonFile } from './fs.js'
-import { hasValue, isArray, isObject } from './util.js'
+import { log } from './log.js'
+import {
+  hasValue,
+  hrPath,
+  isObjectyObject,
+  isPathLike,
+  isString,
+  toPath,
+} from './util.js'
 
 /**
- * @import {LavaMoatPolicy, LavaMoatPolicyOverrides, LavaMoatPolicyDebug} from 'lavamoat-core'
- * @import {WritableFsInterface} from './types.js'
+ * @import {RootPolicy, LavaMoatPolicy, LavaMoatPolicyDebug} from 'lavamoat-core'
+ * @import {LoadPoliciesOptions, WritableFsInterface} from './types.js'
+ * @import {ReadPolicyOptions, ReadPolicyOverrideOptions} from './internal.js'
  */
 
 /**
  * Reads a `policy.json` from disk
  *
- * @param {string | URL} [filepath]
+ * @param {string | URL} policyPath Path to policy
+ * @param {ReadPolicyOptions} [options] Options
  * @returns {Promise<LavaMoatPolicy>}
  * @internal
  */
-export const readPolicy = async (filepath = constants.DEFAULT_POLICY_PATH) => {
-  const allegedPolicy = await readJsonFile(filepath)
-  assertPolicy(allegedPolicy)
-  return allegedPolicy
+export const readPolicy = async (
+  policyPath,
+  { readFile = nodeFs.promises.readFile } = {}
+) => {
+  /** @type {unknown} */
+  let allegedPolicy
+  try {
+    // eslint-disable-next-line @jessie.js/safe-await-separator
+    allegedPolicy = await readJsonFile(policyPath, { readFile })
+  } catch (err) {
+    if (err instanceof SyntaxError) {
+      throw new Error(
+        `Invalid LavaMoat policy at ${hrPath(policyPath)}; failed to parse JSON`,
+        { cause: err }
+      )
+    }
+    if (/** @type {NodeJS.ErrnoException} */ (err).code === 'ENOENT') {
+      throw new Error(
+        `LavaMoat policy file not found at ${hrPath(policyPath)}`,
+        { cause: err }
+      )
+    } else {
+      throw new Error(
+        `Failed to read LavaMoat policy file at ${hrPath(policyPath)}`,
+        {
+          cause: err,
+        }
+      )
+    }
+  }
+  try {
+    assertPolicy(allegedPolicy)
+    return allegedPolicy
+  } catch (err) {
+    throw new Error(
+      `Invalid LavaMoat policy at ${hrPath(policyPath)}; does not match expected schema`,
+      { cause: err }
```

Comment:
We'd need to see what this looks like in the terminal output, I'm not sure how much detail about the actual validation issues is gonna be displayed (because of nesting etc.)

Reply:
it won't show much at all. we have no proper object or schema validation, which (IMO) is better handled by a purpose-built library with a strong UX instead of handrolling such a thing.  this is something I'd like to implement (#733), but it didn't seem of high importance.

Reply:
fwiw, zod v4 beta just got released, which is probably the first thing I'd look at for this.  they claim massive perf improvements